### DISCUSSION TITLE
How to show `message()`s while knitting?

### DISCUSSION THREAD

**Discussion by fkohrt** [ropensci/targets#559] [Help] [answered]
Using Targets Markdown, I can see that **targets** captures all `message()`s (and other conditions). But what would be the appropriate way to show them _during_ the knitting process? Setting the chunk option `message = FALSE` hides them from the resulting document, but doesn't show them while knitting:

``````Rmd
```{r make, message = FALSE, warning = TRUE}
tar_make() # running a target that contains a `message(...)`
```
``````

The background is that I have a longer running simulation in the pipeline that prints progress information with `message()`, but I can't read them while it's still running.

**Comment by wlandau** [comment, marked as answer]
`tar_make()` messages do not get automatically relayed to the R Markdown logs because the pipeline runs in an external R process. (If you were to run `tar_make(callr_function = NULL)`, you would see the messages, but I do not recommend `callr_function = NULL` because of the reproducibility issues). You can see this with a report that runs `callr` without `targets`:

    ---
    title: "trouble relaying callr messages to R Markdown logs"
    output: html_document
    ---
    
    ```{r print_main, message = FALSE}
    message("printed message in the main process")
    ```
    
    ```{r print_external, message = FALSE}
    callr::r(function() message("external printed message"))
    ```

So I recommend a logging strategy that writes to a physical log file instead of the R console. Either that or you could file an issue at https://github.com/rstudio/rmarkdown/issues asking to relay messages from `callr` processes.

**Reply by fkohrt** [reply]
> because of the reproducibility issues

The reproducibility issue you mention would be that `tar_make()` runs in the current session (instead of a new one), right? I render my documents via

```r
xfun::Rscript_call(
    rmarkdown::render,
    list(input = "document.Rmd")
)
```

So should I still be concerned about reproducibility?

> So I recommend a logging strategy that writes to a physical log file instead of the R console.

That's still a good suggestion for cases where I can control the logging, thanks!

**Reply by wlandau** [reply]
Yes, if you are absolutely sure you will always run the report in a fresh new reproducible process with none of the global detritus of any previous session, then callr_function = NULL should not cause any noticeable problems.

**Reply by fkohrt** [reply]
Perfect! As soon as the simulation ran successfully I can also unset the `callr_function` parameter again.

> you could file an issue at https://github.com/rstudio/rmarkdown/issues asking to relay messages from `callr` processes

I still think that this would be an improvement, so I created the issue here: rstudio/rmarkdown#2188

**Reply by fkohrt** [reply]
Following up on the issue: The suggestion was to use the following to log conditions from within **callr** to a file:

```r
callr_function <- function(...) {
    msgfile <- tempfile()
    result <- callr::r(..., stderr = msgfile)
    message(readLines(msgfile))
    return(result)
}
```

Still an advancement because the conditions are written to the file in real time.
